### PR TITLE
http2 upgrade redux

### PR DESCRIPTION
**Pull request by mandarjog**
h2upgrade options in destination rule were not working :-)

This PR fixes the previously unimplemented feature.
It is purely opt-in at present. User needs to set destination rule to h2upgrade.

The user can ensure that it does not interfere with websockets etc.

I have tested this in my cluster.
e2e tests missing.

### CODE REVIEW

File: pilot/pkg/networking/core/v1alpha3/cluster.go
Comment (ramaraochavali) [inline pilot/pkg/networking/core/v1alpha3/cluster.go]: Should this be called in setUpstreamProtocol instead of here so that we set the upstream protocol at a single place based on evaluating all the conditions?

Comment (mandarjog) [reply pilot/pkg/networking/core/v1alpha3/cluster.go]: Selectrrafficpolicy call find port specific connection settings, which I need to determine if upgrade should take place.

**Comment by mandarjog** [comment]
@yxue can you check for effect on sniffing?

**Comment by hzxuzhonghu** [comment]
It does not seem to influence protocol sniffer as it only works on http port.

File: pilot/pkg/networking/core/v1alpha3/cluster.go
Comment (ramaraochavali) [reply pilot/pkg/networking/core/v1alpha3/cluster.go]: I see. I did not realize it is dependent on `SelectTrafficPolicyComponents`. Probably for outbound clusters, we should reconcile `setUpstreamProtocol` logic into `applyTrafficPolicy` and only use `setUpstreamProtocol` for inbound clusters. We can do once this lands.

File: pilot/pkg/networking/core/v1alpha3/cluster.go
Comment (yxue) [inline pilot/pkg/networking/core/v1alpha3/cluster.go]: If the port is unnamed, the function returns false and no h2 options will applied to the cluster. In this case, should the h2 options be applied to the cluster as well since the upstream connection could be h2 and the connection pool will use this h2 options.

Comment (yxue) [reply pilot/pkg/networking/core/v1alpha3/cluster.go]: This seems to be the only place to affect sniffing.

Comment (mandarjog) [reply pilot/pkg/networking/core/v1alpha3/cluster.go]: Thanks

**Comment by mandarjog** [comment]
This is unexpected. I thought that WebSockets tests will fail if I apply h2 upgrades globally @rshriram ?


Did I not change the default in all the required places? @ostromart

**Comment by mandarjog** [comment]
@rshriram ping!

**Comment by rshriram** [comment]
@ramaraochavali can you take a look please?

**Comment by mandarjog** [comment]
@ramaraochavali everything is passing, lets get it in.

**Comment by ramaraochavali** [comment]
LGTM. Does any one want to take another look or should I approve?

**Comment by mandarjog** [comment]
@ramaraochavali I have put a hold on it, please lgtm if you are ok. We will give others another day to weigh in. Thanks.

**Review by howardjohn — APPROVED** [review]
would be nice to get some e2e testing - not sure how easy that is to add though

**Comment by mandarjog** [comment]
@howardjohn I will add e2e in a followup.

**Review by hzxuzhonghu — APPROVED** [review]

**Comment by mandarjog** [comment]
@ostromart when I had a meshconfig snippet in the default values and I ran make_gen it replaces commented meshconfig with an uncommented one. I did not expect that, can you check?
Note: this PR only adds a comment in values now so it is not an issue here.

**Pull request merged**
